share/spack/templates/modules/modulefile.lua

```diff
@@ -71,6 +71,7 @@ depends_on("{{ module }}")
 {% endblock %}
 
 {% block environment %}
+{% set manpath_delim_appended = [] %}
```

Comment:
Hmm, I'd avoid states in the template. Let me think if I came come up with something. :thinking:

Reply:
@alalazo Have you found a better way to avoid states in the template?

Proposed solution here is to have more complex template to get nicer modulefile.

It can be redone to have a simpler template but duplicate colon append in modulefile (which is not an issue for Modules and Lmod).

Reply:
I'll have a look at this again soon :tm: Now that we are past ISC I should have a bit more time.

Reply:
> Proposed solution here is to have more complex template to get nicer modulefile.

It isn't the complexity that scares me, but having a template with a state. I don't think we should do that, templates should just have substitution patterns to be easy to understand (otherwise we'll need to start computing what happened before a substitution in our :exploding_head: ).

Assuming that `environment-modules` can have some arbitrary piece of `tcl` code in them, and that LMod modules can do the same with `lua`, can't we add a snippet that:
1. On activation checks if `MANPATH` exists in the environment, and if its value ends with `:`
2. If not, it does `append_path("MANPATH", ":")`

? Would that work?

Reply:
Understood for template complexity.

> Assuming that environment-modules can have some arbitrary piece of tcl code in them, and that LMod modules can do the same with lua, can't we add a snippet that:
> 1. On activation checks if MANPATH exists in the environment, and if its value ends with :
> 2. If not, it does append_path("MANPATH", ":")
>
> ? Would that work?

I do not get your point.

My goal here is to detect when generating the template that the modulefile being generated changes `MANPATH`, if so add one `append_path("MANPATH", ":")` directive in generated modulefile.

I am not fluent in Jinja (nor Python) and I have issue with the `environment_modifications` structure which is hard to filter within Jinja context.

May a new function be added in `lib/spack/spack/modules/common.py` next to `environment_modifications` to determine that `MANPATH` is changed? This way the test in Jinja context would be a simple `if` condition.

Reply:
> May a new function be added in lib/spack/spack/modules/common.py next to environment_modifications to determine that MANPATH is changed?

Sure, you can start with something like (beware of typos, I didn't check this function for correctness):
```diff
diff --git a/lib/spack/spack/modules/common.py b/lib/spack/spack/modules/common.py
index ef3f751471..006a17a098 100644
--- a/lib/spack/spack/modules/common.py
+++ b/lib/spack/spack/modules/common.py
@@ -40,7 +40,7 @@
 
 import llnl.util.filesystem
 import llnl.util.tty as tty
-from llnl.util.lang import dedupe
+from llnl.util.lang import dedupe, memoized
 
 import spack.build_environment
 import spack.config
@@ -672,6 +672,7 @@ def configure_options(self):
         return None
 
     @tengine.context_property
+    @memoized
     def environment_modifications(self):
         """List of environment modifications to be processed."""
         # Modifications guessed by inspecting the spec prefix
@@ -742,6 +743,18 @@ def environment_modifications(self):
 
         return [(type(x).__name__, x) for x in env if x.name not in exclude]
 
+    @tengine.context_property
+    def has_manpath_modifications(self):
+        for modification_type, cmd in self.environment_modifications:
+            if not isinstance(
+                cmd, (spack.util.environment.PrependPath, spack.util.environment.AppendPath)
+            ):
+                continue
+            if cmd.name == "MANPATH":
+                return True
+        else:
+            return False
+
     @tengine.context_property
     def autoload(self):
         """List of modules that needs to be loaded automatically."""

```